tony-core/src/main/java/com/linkedin/tony/tensorflow/TonySession.java

```diff
@@ -258,11 +259,14 @@ public void onTaskCompleted(String jobName, String jobIndex, int exitCode) {
     // Also note that, we only short circuit when the chief worker failed, not finished.
     // In addition, we support custom jobtype config`tony.application.stop.on.failure.jobtypes`,
     // and we short circuit when that jobtype instance failed.
+    // We added a configuration parameter called `tony.application.fail.on.worker.failure.enabled` with default
+    // value “false” to continue training if some worker jobs failed. If we configured it “true” then it would
+    // immediately stop training if a worker failed.
     if (exitCode != ContainerExitStatus.SUCCESS && exitCode != ContainerExitStatus.KILLED_BY_APPMASTER) {
-      if (isChief(jobName, jobIndex) || shouldStopOnFailure(jobName)) {
+      if (isChief(jobName, jobIndex) || shouldStopOnFailure(jobName) || isFailOnWorkerFailure()) {
         trainingFinished = true;
+        setFinalStatus(FinalApplicationStatus.FAILED, "Exit status: " + exitCode);
```

Comment:
Isn't this `FinalApplicationStatus` a duplicate from the one that `updatesessionStatus()` will give? We should handle `FinalApplicationStatus` only at the session-level but not at the task-level.

Reply:
This was for short circuit the training in case of failure. In updatesessionStatus(),  it first checked the status  and return without scanning each task status.

Reply:
I see, could you add a comment here on the short-circuiting

Reply:
In this case, is it necessary to check `isFailOnWorkerFailure()` again in the `updateSessionStatus()`?

Reply:
Yes, Incase it didn't catch it onTaskCompleted().

Reply:
Could you add some comments on under what case it won't catch it here. Thanks

Reply:
Done